feat(accounts): support configuration under "settings"

mitchnielsen opened the pull request:
## Summary

The [Account API](https://app.prefect.cloud/api/docs#tag/Accounts/operation/update_account_api_accounts__account_id__patch) now has a `settings` block for a few configuration fields. This PR adds support for that new set of fields.

Closes https://github.com/PrefectHQ/terraform-provider-prefect/issues/227

## References

- https://github.com/PrefectHQ/nebula/pull/5454
- [API docs: update account settings](https://app.prefect.cloud/api/docs#tag/Accounts/operation/update_account_settings_api_accounts__account_id__settings_patch)
- [Terraform provider docs: single nested attributes](https://developer.hashicorp.com/terraform/plugin/framework/handling-data/attributes/single-nested)

## Testing

I tested in staging with a personal account:

```terraform
terraform {
  required_providers {
    prefect = {
      source = "registry.terraform.io/prefecthq/prefect"
    }
  }
}

provider "prefect" {
  endpoint     = "https://api.stg.prefect.dev"
  account_id   = "<my account ID>"
  workspace_id = "<my workspace ID>"
  api_key      = "<my API key>"
}

resource "prefect_account" "mitchell-nielsen-testing" {
  name          = "Mitchell Nielsen Testing"
  handle        = "mitchell-nielsen-testing"
  billing_email = "mitchell+testing@prefect.io"

  settings = {
    allow_public_workspaces = true
    ai_log_summaries        = true
    managed_execution       = true
  }
}
```

Before continuing, you'll have to import the account:

```
tf import prefect_account.<account name> <account ID>
```

Once it's imported, adjust each of the `settings` and confirm that they are updated in Prefect:

- AI log summaries and managed execution are in account settings: https://app.stg.prefect.dev/organization/<account ID>/settings
- I'm not really sure where "allow public workspaces" appears in the UI, but the API doesn't complain 😄 

```
$ tf apply -auto-approve

Terraform will perform the following actions:

  # prefect_account.<account ID> will be updated in-place
  ~ resource "prefect_account" "<account ID>" {
        id            = "<account ID>"
        name          = "<account name>"
      ~ settings      = {
          ~ managed_execution       = false -> true
            # (2 unchanged attributes hidden)
        }
      ~ updated       = "2024-07-10T23:56:21Z" -> (known after apply)
        # (3 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
prefect_account.<account name>: Modifying... [id=<account ID>]
prefect_account.<account name>: Modifications complete after 1s [id=<account ID>]

Apply complete! Resources: 0 added, 1 changed, 0 destroyed.

mitchnielsen committed:
Create Settings struct for Accounts

There's a new, nested configuration block for Accounts called Settings.
This adds that struct in the API, data source, and resource.

Closes https://github.com/PrefectHQ/terraform-provider-prefect/issues/227

mitchnielsen committed:
Update Account example with Settings

Adds Settings to the Account example

mitchnielsen commented:
So when I try to update settings against the `/accounts` endpoint, I get a `500 internal server error`. It works when I post directly to the [/accounts/settings endpoint](https://app.prefect.cloud/api/docs#tag/Accounts/operation/update_account_settings_api_accounts__account_id__settings_patch).

I'll dig in more, but unfortunately it looks like you may only be able to update account settings at a specific endpoint. So we'd need to try to calculate what changed in the plan, and hit a different endpoint if Account settings changed 🤔

mitchnielsen committed:
Define a separate UpdateSettings method

When Settings have changed, we need to call a different API endpoint.
This provides a separate Update method called UpdateSettings that
provides the new API endpoint and the slightly different payload.

This probably isn't ideal, and we may want Account Settings to be its
own Terraform resource. But this works for now as an iteration.

github-actions[bot] committed:
Generate Terraform Docs

mitchnielsen commented:
@parkedwards - I'll open this up for your perspective. I left a comment above with my thoughts. I'm comfortable with closing this and creating a dedicated `account_settings` resource if that ends up being the better path, but this implementation does work for me.

I tested manually for now but I'm also happy to update the acceptance test we have for accounts, and expand it to try adjusting these new settings. We currently only confirm an account can be imported, meaning we don't currently test any of the other fields that accounts have.

parkedwards commented on internal/api/accounts.go:
```diff
@@ -8,9 +8,17 @@ import (
 }
 
+// Settings is a representation of an account's settings.
+type Settings struct {
```
i might call this `AccountSettings` to avoid any naming collisions in the future, bc all the structs in the `api` module share scope

mitchnielsen replied:
Great call, changed in c2aa1574c55ea687.

parkedwards commented on internal/provider/datasources/account.go:
```diff
@@ -102,9 +103,26 @@ Use this data source to obtain account-level attributes
-			"allow_public_workspaces": schema.BoolAttribute{
-				Computed:    true,
-				Description: "Whether or not this account allows public workspaces",
+			"settings": schema.SingleNestedAttribute{
+				Description: "Group of settings related to accounts",
```
nice!

parkedwards commented on internal/provider/resources/account.go:
```diff
@@ -214,19 +252,37 @@ func (r *AccountResource) Update(ctx context.Context, req resource.UpdateRequest
+	}
+
+	// If settings have changed, we need to create a separate request to update them.
+	if !plan.Settings.Equal(state.Settings) {
```
nicely done

parkedwards commented:
@mitchnielsen this looks great, i actually like the ergonomics of the `settings` object being embedded in the `prefect_account` resource.  i'd vote for this direction, vs having a separate "settings" resource

there's one thing that i noticed when pulling this down and testing - once you define the `settings` block, it doesn't look like you can "unset" it (eg. remove it) or else we hit a panic. im suspecting something we might need to add something to handle this edge case in the update method.  maybe checking for change, but also if the `settings` block is present?

https://www.loom.com/share/e074b6c5b03441399cd592a1e6c91559?sid=04363e49-27a9-4b7f-ad4a-54407d0f8240

mitchnielsen committed:
Rename 'Settings' to 'AccountSettings'

Renames 'Settings' since all the structs in the 'api' module share
scope. Keeps the 'settings' struct tag since that's how it's represented
in the API itself.

mitchnielsen committed:
Mark "settings" attribute as computed

Marks the "settings" attribute as computed, which will allow Terraform
to return the current settings values from the API if they are not
defined in user configuration.

Also removes the default values, as they're not needed since they'll be
read from the API.

mitchnielsen commented:
@parkedwards great call-out. I pushed d9c8cdc, which seems to do the trick. I was able to change the values and have them applied, and then comment out either one field or the entire `settings` block and the state would be maintained. Let me know what you think.

mitchnielsen committed:
Merge branch 'main' into fix-account-resource-fields
